ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -0,0 +1,414 @@
+            return (epoch, Tip validBlk)
+        _  -- Epoch should be truncated or is empty
+          | epoch > 0 -> go (epoch - 1)
+          | otherwise -> do
+            -- Found no valid blocks on disk
+            removeFilesStartingFrom hasFS 0
+            return (0, TipGen)
+
+-- | Iff the epoch is the most recent epoch, it should not be finalised.
+--
+-- With finalising, we mean: if there are one or more empty slots at the end
+-- of the epoch, the primary index should be padded with offsets to indicate
+-- that these slots are empty. See 'Primary.backfill'.
+data ShouldBeFinalised
+  = ShouldBeFinalised EpochSize
+  | ShouldNotBeFinalised
+  deriving (Show)
+
+-- | Validate the given epoch
+--
+-- * Invalid or missing epoch files will cause truncation. All blocks after a
+--   gap in blocks (due to a missing blocks or invalid block(s)) are
+--   truncated.
+--
+-- * Epoch files are the main source of truth. Primary and secondary index
+--   files can be reconstructed from the epoch files using the
+--   'EpochFileParser'. If index files are missing, corrupt, or do not match
+--   the epoch files, they are overwritten.
+--
+-- * The 'EpochFileParser' checks whether the hashes (header hash) line up
+--   within an epoch. When they do not, we truncate the epoch, including the
+--   block of which its previous hash does not match the hash of the previous
+--   block.
+--
+-- * This function checks whether the first block in the epoch fits on the lat
+--   block of the previous epoch by checking the hashes. If they do not fit,
+--   this epoch is truncated and @()@ is thrown.
+--
+-- * When an invalid block needs to be truncated, trailing empty slots are
+--   also truncated so that the tip of the database will always point to a
+--   valid block or EBB.
+--
+-- * All but the most recent epoch in the database should be finalised, i.e.
+--   padded to the size of the epoch.
+--
+-- TODO currently, we focus on detecting partial writes, not on detecting all
+-- kinds of corruptions, i.e. bitflips. In reality, the likeliness of a
+-- bitflip happening in an epoch file is much larger than one happening in a
+-- secondary index file. So when the checksums don't match, it will probably
+-- be because the block is corrupt, not because the checksum got corrupted in
+-- the secondary index file. We currently assume the epoch file is correct and
+-- overwrite the secondary index file with the new checksum.
+--
+-- In the future, we will be able to check the integrity of a block on its own
+-- so we can determine whether the block or the checksum got corrupted. See
+-- #1253.
+validateEpoch
+  :: forall m hash h e. (IOLike m, Eq hash, HasCallStack)
+  => ValidateEnv m hash h e
+  -> ShouldBeFinalised
+  -> EpochNo
+  -> Maybe (WithOrigin hash)
+     -- ^ The hash of the last block of the previous epoch. 'Nothing' if
+     -- unknown. When this is the first epoch, it should be 'Just Origin'.
+  -> ExceptT () m (Maybe (BlockOrEBB, hash))
```

Comment:
Add a comment about when an error will be thrown (missing epoch file or it doesn't fit onto the previous one). When a block is failing, we don't throw an error, we just truncate and we will fail when validating the next epoch and it doesn't fit onto this one.